mariaDB4j-maven-plugin: exposed database url as Maven Project property

robin-xyzt-ai opened the pull request:
There is currently no way to know on which port the `mariaDB4j-maven-plugin` has started the database so that integration tests can connect to it.
The only option is to hardcode a port number in the configuration of the `mariaDB4j-maven-plugin`, and use that same hardcoded port number in your test.

The old plugin version from mike10004 exposed a port number as Maven Project property (see [this file](https://github.com/mike10004/mariadb4j-maven-plugin/blob/master/mariadb4j-maven-plugin/src/main/java/com/github/mike10004/mariadb4jmavenplugin/Mariadb4jFindFreePortMojo.java)).

This PR does more or less the same. Instead of exposing the port number, it exposes the database URL. I also updated the readme to show how to retrieve that database URL and pass it to your integration tests as system property.

vorburger reviewed (approved):
LGTM! @robin-xyzt-ai thank you for your contribution; please keep it coming if you have others.

vorburger reviewed (changes requested):
@robin-xyzt-ai actually... sorry, this does LGTM - BUT it's breaking the tests! I'm getting a NPE `at ch.vorburger.mariadb4j.StartMojo.runWithMavenJvm(StartMojo.java:64)`. If this worked for you when you wrote it, then perhaps it broke when I merged #477 or #478 or #482 or #483 today? If you want to see this merged, could you rebase and fix whatever is causing that NPE?

vorburger commented:
#489 doesn't seem to help to fix the NPE.

vorburger commented:
@robin-xyzt-ai BTW it would be cool to have "test coverage" for this - do you want to add what you are describing in the README to the https://github.com/vorburger/MariaDB4j/blob/master/mariaDB4j-maven-plugin/src/it/mariadb4j-maven-plugin-test-basic/pom.xml and then read that system property in https://github.com/vorburger/MariaDB4j/blob/master/mariaDB4j-maven-plugin/src/it/mariadb4j-maven-plugin-test-basic/src/test/java/ch/vorburger/mariadb4j/mariadb4jmavenplugintest/basic/BasicUsageIT.java?

robin-xyzt-ai commented:
I wanted to write tests for this, but at the time of this PR the build was failing so I couldn't run the tests through Maven. Perhaps I should have set up a real project and do it in there. I will have another look at this

robin-xyzt-ai commented:
Added a commit which fixes the NPE. Fixed it by switching from `lookupMojo` to `lookupConfiguredMojo` in the tests, based on the comments underneath [this blog post](https://vzurczak.wordpress.com/2014/07/23/write-unit-tests-for-a-maven-plug-in/).

I admit that I don't know if this change has any other unwanted effects. It's my first time writing/testing a Maven plugin and the documentation on testing a Maven plugin is a bit lacking. E.g. those methods don't even have decent Javadoc telling me what the difference is between both.

robin-xyzt-ai commented:
@vorburger : can you tell me how I can run `BasicUsageIT.java` test from Maven.

Currently I'm using the same as what travis is using (`mvn verify -B -V` in the root directory), but that doesn't seem to trigger that `BasicUsageIT.java` test. I could not find it in the Maven output, and when I tried to simply throw an exception in that test the build is still a success.

That makes it a bit hard to test whether my changes to the test have any effect before I commit them.

vorburger commented:
> @vorburger : can you tell me how I can run `BasicUsageIT.java` test from Maven.
> 
> Currently I'm using the same as what travis is using (`mvn verify -B -V` in the root directory), but that doesn't seem to trigger that `BasicUsageIT.java` test. I could not find it in the Maven output, and when I tried to simply throw an exception in that test the build is still a success.

Oh! I actually never noticed that myself...

@duttonw who originally contributed this in #135 may want to chime in here?

You could also "manually" test it - I wouldn't be against merging this as-is, if you want me to.

robin-xyzt-ai commented:
I think I've done everything you asked for this PR:

- The NullPointerExceptions in the test are fixed
- Rebased on the current version of master for easier merging
- Added an assertion to the BasicUsageIT (but since I don't know how to run it, I couldn't verify it actually works)

robin-xyzt-ai commented:
Thank you for accepting and merging this PR.
